cmds/exp/dmidecode/dmidecode.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2016-2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	flag "github.com/spf13/pflag"
+
+	"github.com/u-root/u-root/pkg/smbios"
+)
+
+var (
+	flagFromDump = flag.String("from-dump", "", `Read the DMI data from a binary file previously generated using --dump-bin.`)
+	flagType     = flag.StringSliceP("type", "t", nil, `Only  display  the  entries of type TYPE. TYPE can be either a DMI type number, or a comma-separated list of type numbers, or a keyword from the following list: bios, system, baseboard, chassis, processor, memory, cache, connector, slot. If this option is used more than once, the set of displayed entries will be the union of all the given types. If TYPE is not provided or not valid, a list of all valid keywords is printed and dmidecode exits with an error.`)
+	// NB: When adding flags, update resetFlags in dmidecode_test.
+)
+
+var (
+	typeGroups = map[string][]uint8{
+		"bios":      {0, 13},
+		"system":    {1, 12, 15, 23, 32},
+		"baseboard": {2, 10, 41},
+		"chassis":   {3},
+		"processor": {4},
+		"memory":    {5, 6, 16, 17},
+		"cache":     {7},
+		"connector": {8},
+		"slot":      {9},
+	}
+)
+
+// parseTypeFilter parses the --type argument(s) and returns a set of types taht should be included.
+func parseTypeFilter(typeStrings []string) (map[smbios.TableType]bool, error) {
+	types := map[smbios.TableType]bool{}
+	for _, ts := range typeStrings {
+		if tg, ok := typeGroups[strings.ToLower(ts)]; ok {
+			for _, t := range tg {
+				types[smbios.TableType(t)] = true
+			}
+		} else {
+			u, err := strconv.ParseUint(ts, 0, 8)
+			if err != nil {
+				return nil, fmt.Errorf("Invalid type: %s", ts)
+			}
+			types[smbios.TableType(uint8(u))] = true
+		}
+	}
+	return types, nil
+}
+
+func dmiDecode(textOut io.Writer) (int, error) {
```

Comment:
int is the exit code in case of error

Reply:
so you might want to look at ways to do this other than multiple value return.
https://blog.golang.org/error-handling-and-go

Reply:
i guess the code could be folded into a special error type, then extract it from there... however, since it's a function with exactly one call site, that is local to the file and is not part of an interface, seems like more work than it's worth.